common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java

```diff
@@ -353,10 +354,11 @@ public static String toUpperCase(final String target) {
    * @return the uppercase string
    */
   public static UTF8String toUpperCase(final UTF8String target, final int collationId) {
-    return UTF8String.fromString(toUpperCase(target.toString(), collationId));
+    if (target.isFullAscii()) return target.toUpperCaseAscii();
+    return UTF8String.fromString(toUpperCaseSlow(target.toString(), collationId));
   }
 
-  public static String toUpperCase(final String target, final int collationId) {
+  private static String toUpperCaseSlow(final String target, final int collationId) {
```

Comment:
Same thing about passing and returning `UTF8String` values here.

Reply:
discussing in: https://github.com/apache/spark/pull/47014#discussion_r1652818328